model/sharing/sharing.go

```diff
@@ -1056,206 +1058,592 @@ func CheckSharings(inst *instance.Instance) ([]map[string]interface{}, error) {
 			}
 		}
 
-		// Check triggers
-		if s.Active && accepted {
-			if s.Triggers.TrackID == "" && len(s.Triggers.TrackIDs) == 0 {
+		checks, validMembers := s.checkSharingMembers()
+		checks = append(checks, s.checkSharingTriggers(inst, accepted)...)
+		checks = append(checks, s.checkSharingCredentials()...)
+
+		if len(checks) == 0 {
+			if !s.Owner {
+				return nil
+			}
+
+			parentSharing, err := getParentFileSharing(inst, s)
+			if err != nil {
+				return err
+			} else if parentSharing != nil {
 				checks = append(checks, map[string]interface{}{
-					"id":      s.SID,
-					"type":    "missing_trigger_on_active_sharing",
-					"trigger": "track",
+					"id":             s.SID,
+					"type":           "sharing_in_sharing",
+					"parent_sharing": parentSharing.SID,
+				})
+				return nil
+			}
+
+			if !s.Active || s.Initial || s.ReadOnly() {
+				return nil
+			}
+
+			isSharingReady := false
+			for _, m := range s.Members {
+				if m.Status == MemberStatusReady {
+					isSharingReady = true
+					break
+				}
+			}
+			if !isSharingReady {
+				return nil
+			}
+
+			rule := s.FirstFilesRule()
+			if rule == nil {
+				return nil
+			}
+
+			ownerDocs, err := FindMatchingDocs(inst, *rule)
+			if err != nil {
+				checks = append(checks, map[string]interface{}{
+					"id":    s.SID,
+					"type":  "missing_matching_docs_for_owner",
+					"error": err.Error(),
 				})
-			} else if s.Triggers.TrackID != "" {
-				err := couchdb.GetDoc(inst, consts.Triggers, s.Triggers.TrackID, nil)
+				return nil
+			}
+
+			var freeOwnerDiskSpace int64
+			if usage, err := inst.VFS().DiskUsage(); err != nil {
+				freeOwnerDiskSpace = inst.VFS().DiskQuota() - usage
+			}
+
+			for _, m := range validMembers {
+				ms, err := FindSharing(m, s.ID())
+				if err != nil {
+					checks = append(checks, map[string]interface{}{
+						"id":    s.SID,
+						"type":  "missing_sharing_for_member",
+						"error": err.Error(),
+					})
+					continue
+				}
+
+				parentSharing, err := getParentFileSharing(m, ms)
+				if err != nil {
+					return err
+				} else if parentSharing != nil {
+					checks = append(checks, map[string]interface{}{
+						"id":             ms.SID,
+						"type":           "sharing_in_sharing",
+						"parent_sharing": parentSharing.SID,
+					})
+					continue
+				}
+
+				if !skipFSConsistency {
+					checks = append(checks, s.checkSharingTreesConsistency(inst, ownerDocs, freeOwnerDiskSpace, m, ms)...)
+				}
+			}
+		}
+
+		return nil
+	})
+	return checks, err
+}
+
+// getParentFileSharing returns the first sharing found accepting the root of
+// the given sharing.
+//
+// Since have a sharing within another one will generate unexpected behavior,
+// the goal is to find these situations.
+func getParentFileSharing(inst *instance.Instance, sharing *Sharing) (*Sharing, error) {
+	if sharing.DocType() != consts.Files {
+		return nil, nil
```

Comment:
I don't like functions that can return `nil, nil`. I prefer to add a third return parameter like `(*Sharing, bool, error)` and `return nil, false, nil` or `return something, true, nil`.

I know when reading code, I make easily the assumption that after a block like:

```go
foo, err := doSomething()
if err != nil {
    return err
}
```

that `foo` is not nil. And it was the cause of a few bugs that could have been avoided with the third return parameter trick.

Reply:
Is it the same reason why the linter allows me to return only `nil` from `CheckSharings()` while it's expecting a tuple as output?

Reply:
No, the compiler doesn't allow that. There are no `return nil` for `CheckSharings`, only for the anonymous function passed as a parameter to `couchdb.ForEach` inside `CheckSharings`.

Reply:
Right. I read a bit fast.

Reply:
So you'd write something like this?

```go
parentSharing, ok, err := getParentFileSharing(m, ms)
if err != nil {
	return err
} else if ok {
	checks = append(checks, map[string]interface{}{
		"id":             ms.SID,
		"type":           "sharing_in_sharing",
		"parent_sharing": parentSharing.SID,
	})
	continue
}
```

Reply:
Yes